### PR TITLE
feat: add a `transform` method

### LINKED ISSUE #107: [Feature] middleware / custom transforms support
currently, what is returned after crawling and what is used in filters is always what fdir has processed. i propose a function that allows users to transform those paths with custom logic.

why? because if a user wants to use their custom logic to transform fdir paths they currently have to duplicate it if using filters/excludes/globs (which also means the transform logic is ran twice, which is bad for performance):

```js
const files = await new fdir()
  .withRelativePaths()
  .filter(p => {
    // whatever custom logic they might need
    const path = `../${p}`;

    return someFilteringFunction(path);
  })
  .crawl(root)
  .withPromise()

// the exact same custom logic again
return files.map(p => `../${p}`)
```

with support for custom transforms, it could look something more like this:

```js
return new fdir()
  .withRelativePaths()
  .transform(p => `../${p}`)
  .filter(someFilteringFunction)
  .crawl(root)
  .withPromise()

// it returns paths with a leading `../`, without duplicating the transform logic. woohoo!
```

see a real world example of the problem this proposal aims to solve in https://github.com/SuperchupuDev/tinyglobby/blob/1b51f7b90fb0b7db09d6d656b6f80341a0eaeffd/src/index.ts#L72-L73 and https://github.com/SuperchupuDev/tinyglobby/pull/18/files' `processPath` function

### PR DESCRIPTION
**Pull request by ziebam**
Resolves #107.

The issue and the code say everything, really. 😄

### CODE REVIEW

**Review by SuperchupuDev — CHANGES_REQUESTED** [review]
as per the discord conversation, can you add a `transform` argument to the filter function?

**Comment by ziebam** [comment]
> as per the discord conversation, can you add a `transform` argument to the filter function? or some other way of being able to transform before filtering at all

For simplicity's sake, I think supporting just one or the other is better. Based on your use cases transforming before the filtering makes more sense. Pushed the change and clarified that in the documentation.

File: src/api/functions/push-directory.ts [hunk 1]
```diff
@@ -1,39 +1,50 @@
...
+  filters?: FilterPredicate[],
+  transformer?: TransformPredicate
 ) => void;
 
 function pushDirectoryWithRelativePath(root: string): PushDirectoryFunction {
-  return function (directoryPath, paths) {
-    paths.push((directoryPath || ".").substring(root.length));
+  return function (directoryPath, paths, _filters, transformer) {
+    const relativePath = (directoryPath || ".").substring(root.length);
+    paths.push(transformer?.(relativePath, true) ?? relativePath);
```
Comment (SuperchupuDev) [inline src/api/functions/push-directory.ts:13]: optional chaining would sadly break compatibility with anything above node 14

**Review by SuperchupuDev — CHANGES_REQUESTED** [review]
also pending the second argument change to a `Dirent`, just leaving it here as a reminder, amazing work so far :-)

File: src/api/functions/push-directory.ts (hunk 1 above)
Comment (ziebam) [reply src/api/functions/push-directory.ts:13]: Oh my, got used to those features so much at work that I forgot they weren't supported at one point. Gonna change later.

**Comment by thecodrr** [comment]
Hey, I really appreciate the effort but as I mentioned in #107, fdir isn't the right place for this.

P.S. please discuss before implementing any feature to avoid situations like these.

**Comment by SuperchupuDev** [comment]
is there any place we can contact you outside github to discuss fdir related stuff? we've been discussing our fdir prs in the [e18e discord](https://chat.e18e.dev/) which you might be interested in joining

**Comment by thecodrr** [comment]
I think the discussion should happen under a GitHub issue as that keeps things more localized and easier to reference later on.

With that said, let me know if you prefer something more...informal. I can create a discord server for fdir but to be honest, I don't think the world needs yet another discord server. There are already too many. 😂

**Pull request closed without merging**